gennet-module/gennet.py

```diff
@@ -256,33 +287,79 @@ def dict_to_arrays(dic):
     return keys, vals
 
 
+# Check for range failures in a list
+def range_fails(lst, min=0, max=100):
+    return any(x < min or x > max for x in lst)
+
+
+# Check for sum failures in a list
+def sum_fails(lst, sum_expected=100):
+    return not sum(lst) == sum_expected
+
+
+# Construct the nodeType from the trait
+def traits_to_nodeType(s):
+    return nodeType(s.split(':')[0])
+
+# Validate the traits distribution (stick to percentages: num nodes may vary post generation)
+def validate_traits_distribution(traits_distribution, num_nodes):
+    traits, traits_freq = dict_to_arrays(traits_distribution)
+    if range_fails(traits_freq, max=100) :
+        raise ValueError(f"{traits_distribution} : invalid percentage (>{100} or <0)")
+    if sum_fails(traits_freq, sum_expected=100) :
+        raise ValueError(
+                f"{traits_distribution} : percentages do not sum to {100}")
+    for s in traits :
```

Comment:
This can be better organized as:

```python
all_traits = set([trait for trait_string in traits for trait in trait_string.split(":")])
for trait in all_traits:
    if trait not in Trait and not os.path.exists(f"{TRAITS_DIR}/{t}.toml") :
        raise ValueError(f"{traits_distribution} : unknown trait {t} in {s}")
    print(f"{TRAITS_DIR}/{t}.toml")
```

Reply:
writing it nested helps better error reporting: the ValueError clearly accounts which trait {t} from specific trait string {s}  is the issue. I first did a very concise `any(..)`, but abandoned it because `any(..)` doesn't return the value for which it fails.

Reply:
I had to modify the `raise` cause indeed you don't have access to the same values, but imo:
1 - This comprehension is more readable, compact and faster.
2 - Doing the set get rid of repeated values, avoiding unnecessary checkings.
3 - It is a bit unnecessary to check where the trait is, as each trait is supposed to be a file. Just pointint out which file is, should be enough.

Reply:
Set will internally have to iterate and de-duplicate anyway. comprehension and iteration means we will be looping over the lists twice! on the traits, Agree checking for files alone is ok, but I finally  want to populate the "enum" with contents of traits dir (config.json might specify the trait dir) to make everything parameterisable in Gennet. so let it stay this way for now.

Reply:
Also, we should agree on do not close conversations until both parts agree.

Example, I tell you to get rid of an unnused variable. If you do it, then ofc you can close the conversation because both parts agreed. Here, agreemend is still not made so we can keep talking about this.

First, sets consumes by hash, it doesnt dedup by double iterating. 
Also, ofc a list comprehension iterates, but as I mentioned, it is much faster and also compact.

Reply:
Maybe coming late to the party but IMO the issue is repeating this:

```python
if t not in Trait and not os.path.exists(f"{TRAITS_DIR}/{t}.toml") :
                raise ValueError(
                    f"{traits_distribution} : unknown trait {t} in {s}")
```
for the duplicated tags. Usually an OS/IO call will be expensive and python sets/dictionaries are pretty fast.

Reply:
That's a fair point, but two things: OS/IO calls are only for user defined traits (which I suspect are rather rare) and python logical operators 'short circuit'.

Reply:
Well, I didn't check what Traits is though. If you are sure should be ok I guess. But it would not do any harm anyway 😅

Reply:
Traits are basically my fancy way of calling different wakunode capabilities; for instance "store", "relay", "discv5", "dnsdiscv" , "relay", "filter"  etc are traits; each trait has an associated toml file and we build the toml for a node from its trait string. for instance, "nwaku:discv5:store" will launch a nwaku node with store and discv5 capabilities.

the user defined traits for traits that a wakunode can have in the future! :D

Reply:
![image](https://user-images.githubusercontent.com/23367096/227167083-8f79199b-a7d5-449d-b58d-7850d1dffa15.png)

Times for both executions:
Test1: 4.671671699999327
Test2: 0.7970336666621733

Reply:
First, WHY would you want to specify ["nwaku:store:discv5":0.2, "nwaku:sore:discv5":0.2, ... 500 times] ?  the way to specify that  (that all nodes follow the same trait string) is as follows  ["nwaku:store:discv5":100].

And then the cherry picking:  duplicating a string 500 times will *obviously* favour set; the same way how no-duplicates will favour the code in repo. What are you trying to say here?

Reply:
I do appreciate the feedback ofc. But this one is complicated on many levels. For instance, remember, the json parser already *de-duplicates* at the trait string level. so the situation 500 fold duplicate like the one mentioned will simply NOT arise:  trait string is a key in a dict, it can be present at most once.

The trait distributions themselves are fairly compact and most  probably consists of less than 10 trait strings. And this validation runs just once at the beginning.

I am all for improving the code as long it improves the clarity and performance.